### PR TITLE
Minimal Viable Port of the SMG class

### PR DESCRIPTION
**Pull request by petr-muller**
Adds a C++ port of the original SMG class. Just several original methods were ported: this PR serves mainly as a base for further work.

### CODE REVIEW

**Comment by michkot** [comment]
Finding no.1: usage of `__attribute__((unused))` is GCC(clang)-specific, breaking compatibility with other compilers

**Comment by michkot** [comment]
Finding no.2: declaring `object` as `private` in `SMGEdge` and later retrieving it with `getObject`in derived classed leads to copy-constructing the `shared_ptr` upon return of `getObject` and thus making unnecessary overhead (IMHO).

**Comment by michkot** [comment]
Finding no.3: SMGObject.cc:10 `
+const SMGObjectPtr SMGNullObject::NULL_OBJECT = SMGObjectPtr(new SMGNullObject());`
could this also be replaced with `std::make_shared<SMGNullObject>(.....)` to keep consistency with the rest(newer commits) of the code?

**Comment by michkot** [comment]
Attaching no.1 patch is wanted - I finished reviewing the code without any more findings.
[0001-remove-**attribute**-replace-with-void-decoy.zip](https://github.com/VeriFIT/smg3/files/130008/0001-remove-__attribute__-replace-with-void-decoy.zip)

**Comment by petr-muller** [comment]
1. IMO, not a problem, I only care about GCC. Is smg3 really supposed to be compiled with different compilers? IMO, decoys are an anti-pattern: they obfuscate real meaning. I want to declare it's okay that a certain value is not used, not fake its usage. Either we hide it behind a macro or we just comment out the param name.
2. Accessing parent members (and avoiding protected members) via accessors is a Java best practice (it allows you to change the parent class without changing derived ones, reinforcing encapsulation). I do not know C++ best practices well enough to judge. Personally, I prefer encapsulation over performance, unless we have a measurement that it really hurts performance.
3. Actually, I thought I'll get rid of std::make_shared calls (because from the client perspective, its only a happy coincidence that make_shared returns something compatible with SMGObjectPtr. I think I should really provide something like SMGObject::make_shared that returns SMGObjectPtr.

**Comment by petr-muller** [comment]
@michkot Consider using the GH code comment system - I find it easier to discuss code while looking at it :)

**Comment by michkot** [comment]
@petr-muller Ok, sure, will do next time (I was not reviewing the code in browser so...)

File: src/objects/SMGObject.cc [hunk 1]
```diff
@@ -1,42 +1,72 @@
 #include "SMGObject.hh"
 
-long SMGObject::id_counter = 0;
+long SMGObject::id_counter = 1;
+const SMGObjectPtr SMGNullObject::NULL_OBJECT = SMGObjectPtr(new SMGNullObject());
```
Comment (michkot) [inline src/objects/SMGObject.cc:4]: not in union with other occurrences; 2+ memory allocations - inefficient and possibly prone to allocation failures

**Comment by michkot** [comment]
1. I would not say it is primarily supposed to be compiled with different compilers, but, it would be silly to attempt such a cheap "vendor lock-in". My first taught was a macro (but they are generally discouraged), but omitting the name would also do the job. 
   I personally like [this macro](http://stackoverflow.com/a/1670083) , downside is that involving a macro would mean creating and using some kind of "platform.hh" (that's what I call it in my C projects).
2. I get the encapsulation idea - would like to see PP's opinion on this. Problem is, that getObject is not a simple accessor from C++ point of view ;)
3. well... that's the problem with the `typedef`obfuscation. It would be otherwise pretty clear that make_shared<SmgObject> returns shared_ptr<SmgObject>. But I think I got that it was intentional obfuscation, to allow for later replacement of shared_ptr with something else - but it still probably would not be "drop-down replacement". Also, SMGObject::make_shared would then be some kind of SMGObject "factory" and you wold need to implement it for every variant of the SMGObject (Region...) constructor (I hope I am right on this one). Using the name "make_shared" would not make any sense, because it beats the advantage of obfuscating the pointer type via SMGObjectPtr. For me, the question is - will "client" need to manually create a single entity without doing anything else with it immediately after that? If yes, then such a method might/function might be useful, otherwise not. To sum up: 
   a) IMHO do not use shared_ptr<...>(new ...) construct 
   b) Either stick with the `typedef`only for writing convenience and use std::make_shared or we should came with complex strategy for that.

**Comment by michkot** [comment]
Discussed the matter with PP today, TL;DR; version:
1. Omitting the parameter name would probably be the right way
2. To be considered later, during the review when the library is more or less ported.
3. The public interface generally should not contain any "pointers to objects" etc. at all. Typedefs are "eeeehhh". std::make_shared itself is "a factory" and it is the best way when allocating new SMG object + shared_ptr in the implementation.

**Comment by petr-muller** [comment]
1. Ack
2. Ack

The public interface should not ..., _because_...? What is the alternative? Right now, most of the getters that will provide pointers to objects.
typedefs. Understood. What is the alternative? Using the full name everywhere? I prefer having shorter and more descriptive names.

**Pull request merged**